File: src/data/codes/knn.m
clc;  % KNN
clear;
close all;

%% =========================
% SETTINGS
% ==========================
useSynthetic = true;
csvFile = 'knn_student_data.csv';
K = 5;   % change this to study effect of K

%% =========================
% LOAD OR CREATE DATA
% ==========================
if useSynthetic
    rng(10);
    n = 150;

    Study_Hours = randi([1 10], n, 1);
    Attendance = randi([50 100], n, 1);
    Assignment_Score = randi([40 100], n, 1);
    Internal_Marks = randi([10 50], n, 1);

    score = 0.8*Study_Hours + 0.03*Attendance + 0.03*Assignment_Score + 0.08*Internal_Marks;
    Result = double(score > 8.5);

    data = table(Study_Hours, Attendance, Assignment_Score, Internal_Marks, Result);
else
    data = readtable(csvFile);
    % Expected columns:
    % Study_Hours, Attendance, Assignment_Score, Internal_Marks, Result
end

disp(head(data));

%% =========================
% PREPROCESSING
% ==========================
data = rmmissing(data);
data = unique(data);

X = [data.Study_Hours, data.Attendance, data.Assignment_Score, data.Internal_Marks];
Y = data.Result;

X = normalize(X);

%% =========================
% TRAIN-TEST SPLIT
% ==========================
cv = cvpartition(Y, 'HoldOut', 0.2);
X_train = X(training(cv), :);
Y_train = Y(training(cv), :);
X_test  = X(test(cv), :);
Y_test  = Y(test(cv), :);

%% =========================
% MODEL BUILDING
% ==========================
model = fitcknn(X_train, Y_train, 'NumNeighbors', K);

%% =========================
% PREDICTION
% ==========================
Y_pred = predict(model, X_test);

%% =========================
% EVALUATION
% ==========================
confMat = confusionmat(Y_test, Y_pred);
accuracy = sum(diag(confMat)) / sum(confMat(:));

TP = confMat(2,2); TN = confMat(1,1);
FP = confMat(1,2); FN = confMat(2,1);

precision = TP / (TP + FP + eps);
recall    = TP / (TP + FN + eps);
f1        = 2 * precision * recall / (precision + recall + eps);

disp('Confusion Matrix:');
disp(confMat);
fprintf('\nK = %d\n', K);
fprintf('Accuracy  = %.4f\n', accuracy);
fprintf('Precision = %.4f\n', precision);
fprintf('Recall    = %.4f\n', recall);
fprintf('F1-score  = %.4f\n', f1);

%% =========================
% EFFECT OF K VALUE
% ==========================
k_values = 1:10;
acc_values = zeros(size(k_values));

for i = 1:length(k_values)
    tempModel = fitcknn(X_train, Y_train, 'NumNeighbors', k_values(i));
    tempPred = predict(tempModel, X_test);
    tempConf = confusionmat(Y_test, tempPred);
    acc_values(i) = sum(diag(tempConf)) / sum(tempConf(:));
end

figure;
plot(k_values, acc_values, 'o-', 'LineWidth', 1.5);
xlabel('K value');
ylabel('Accuracy');
title('Effect of K on KNN Accuracy');
grid on;

figure;
confusionchart(Y_test, Y_pred);
title('KNN Confusion Matrix');

File: src/data/codes/knn.py
"""KNN classifier
# pip install numpy pandas scikit-learn matplotlib
"""
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import ConfusionMatrixDisplay
import matplotlib.pyplot as plt

USE_SYNTHETIC = True
K = 5
RNG = np.random.default_rng(10)

if USE_SYNTHETIC:
    n = 150
    study = RNG.integers(1, 11, size=n)
    att = RNG.integers(50, 101, size=n)
    assign = RNG.integers(40, 101, size=n)
    internal = RNG.integers(10, 51, size=n)
    score = (
        0.8 * study
        + 0.03 * att
        + 0.03 * assign
        + 0.08 * internal
    )
    result = (score > 8.5).astype(int)
    data = pd.DataFrame(
        {
            "Study_Hours": study,
            "Attendance": att,
            "Assignment_Score": assign,
            "Internal_Marks": internal,
            "Result": result,
        }
    )
else:
    data = pd.read_csv("knn_student_data.csv")

print(data.head())
data = data.dropna().drop_duplicates()
X = data[["Study_Hours", "Attendance", "Assignment_Score", "Internal_Marks"]].values
y = data["Result"].values
scaler = StandardScaler()
X = scaler.fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)

model = KNeighborsClassifier(n_neighbors=K).fit(X_train, y_train)
y_pred = model.predict(X_test)

cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", cm)
print(f"\nK = {K}")
print(f"Accuracy  = {accuracy_score(y_test, y_pred):.4f}")
print(f"Precision = {precision_score(y_test, y_pred, zero_division=0):.4f}")
print(f"Recall    = {recall_score(y_test, y_pred, zero_division=0):.4f}")
print(f"F1-score  = {f1_score(y_test, y_pred, zero_division=0):.4f}")

k_values = range(1, 11)
acc_vals = []
for k in k_values:
    m = KNeighborsClassifier(n_neighbors=k).fit(X_train, y_train)
    pred = m.predict(X_test)
    acc_vals.append(accuracy_score(y_test, pred))

fig, ax = plt.subplots()
ax.plot(k_values, acc_vals, "o-", linewidth=1.5)
ax.set_xlabel("K value")
ax.set_ylabel("Accuracy")
ax.set_title("Effect of K on KNN Accuracy")
ax.grid(True)

fig2, ax2 = plt.subplots()
ConfusionMatrixDisplay.from_predictions(y_test, y_pred, ax=ax2)
ax2.set_title("KNN Confusion Matrix")
plt.show()
